### PR TITLE
feat: Posted/edited/closed alert banner

### PR DESCRIPTION
**Pull request by cmaddox5**
**Asana task**: [[Screenplay] Update Banner](https://app.asana.com/0/1185117109217413/1203279317437018/f)

Added a banner to the top of every page showing if an alert has been posted/edited/closed in the last two minutes. Only triggers for significant alerts (see `util.ts` for how we choose significant alerts).

### CODE REVIEW

File: assets/js/components/Dashboard/Dashboard.tsx [hunk 1]
```diff
@@ -6,45 +6,134 @@ import { Alert } from "../../models/alert";
...
+          moment(postedOrEditedAlert.updated_at).isAfter(
+            moment(bannerAlert.closedAt)
+          )
+      )
+    ) {
+      setBannerAlert({ alert: postedOrEditedAlert });
+    }
+    // If no other alert is eligible to be the bannerAlert, expire the current bannerAlert if two minutes have passed since it started displaying.
+    else if (
+      moment(now).isSameOrAfter(moment(bannerAlert.closedAt).add(2, "minutes"))
```
Comment (PaulJKim) [inline assets/js/components/Dashboard/Dashboard.tsx:85]: Would this check handle posted/edited alerts? Since those alerts wouldn't have a defined `closedAt` value according to the comment on lines 65-67

File: assets/js/components/Dashboard/AlertBanner.tsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
+import React, { ComponentType } from "react";
+import { Alert } from "../../models/alert";
+import { ArrowRepeat } from "react-bootstrap-icons";
+import { formatEffect, translateRouteID } from "../../util";
+import { useParams } from "react-router-dom";
+
+interface Props {
+  alert: Alert;
+  isClosedAlert: boolean;
```
Comment (PaulJKim) [inline assets/js/components/Dashboard/AlertBanner.tsx:9]: Definitely a nit but does this need to be prop? We could just do the check on line 39 that we do on line 134 in Dashboard.tsx, I think

File: assets/js/util.ts [hunk 3]
```diff
@@ -46,6 +46,65 @@ export const abbreviation = (line: string) => {
...
+      return "Orange Line";
+    case "blue":
+      return "Blue Line";
+    case "silver":
+      return "Silver Line";
+    case "green":
+    case "green-b":
+    case "green-c":
+    case "green-d":
+    case "green-e":
```
Comment (PaulJKim) [inline assets/js/util.ts:65]: hmm I'm not sure how much of this is specified in the spec, but would it be worth including strings for the different green branches? I think it's useful to distinguish when an alert is affecting just one branch on not other on the green line trunk, for example

File: assets/js/util.ts [hunk 4]
```diff
@@ -46,6 +46,65 @@ export const abbreviation = (line: string) => {
...
+    case "bus":
+      return "Bus";
+    case "cr":
+      return "Commuter Rail";
+    default:
+      return "";
+  }
+};
+
+export const getModeFromAffectedList = (affectedList: string[]) => {
```
[the remark this answers is not in the document]
Comment (cmaddox5) [reply assets/js/util.ts:76]: Hmm, that's a good question. I think that normally alerts only affect one mode, but I'll need to check with product on that.

File: assets/js/components/Dashboard/Dashboard.tsx (hunk 1 above)
Comment (cmaddox5) [reply assets/js/components/Dashboard/Dashboard.tsx:85]: That scenario should not get through the first `else if` on line 68. If `bannerAlert` is `undefined` or `bannerAlert.closedAt` is `undefined` and there is no new closed alert, it will set the `bannerAlert` to the result of `getPostedOrEditedAlert`. That result can either be a posted/edited alert from the last two minutes or `undefined` if there is no recent alert. 

After not looking at this for a week, this conditional block is hard to follow, but I'm really not sure how to simplify it 😬 . If you have any suggestions, let me know.

File: assets/js/util.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (cmaddox5) [reply assets/js/util.ts:76]: Had a talk with PaulH about this in Slack ([thread](https://mbta.slack.com/archives/C039ARUM48H/p1669821605654029)). An alert created for subway _could_ also affect bus (i.e. Station Closure), but not the other way around. And the current logic tries to match on subway first, so this logic works for our needs. As a test, PaulH created a Station Closure alert at Harvard and a banner was displayed for it.

File: assets/js/components/Dashboard/Dashboard.tsx (hunk 1 above)
Comment (PaulJKim) [reply assets/js/components/Dashboard/Dashboard.tsx:85]: Hmm yeah I've been looking at it and can't think of any ways to make any significant simplifications. One small suggestion that might help is putting the `moment` comparisons on Lines 74-77 in a separate helper function to help the conditional block read a bit better, but I think that's the best I can come up with at the `moment` 🤷 

One thing I'm a little confused on is how a posted/edited alert banner will be closed after two minutes. I see that for a closed alert, we're constantly checking if two minutes has passed since the `closedAt` time, but a posted or edited alert wouldn't have that field, right? Oh, or does that check implicitly happen in `getPostedOrEditedAlert` and then the banner alert would just get set as `undefined` on line 69?

Comment (cmaddox5) [reply assets/js/components/Dashboard/Dashboard.tsx:85]: Yep, it happens in `getPostedOrEditedAlert`. For recently posted/edited alerts, we are able to check if `created_at` or `updated_at` is in the last two minutes. Because the alert itself has that data, we don't need to store it and constantly check against the time we first saw it. For a closed alert, those fields are unreliable and would be lost on the second refresh after the alert closed anyway.

File: assets/js/components/Dashboard/AlertBanner.tsx (hunk 2 above)
Comment (cmaddox5) [reply assets/js/components/Dashboard/AlertBanner.tsx]: Good call. I updated the logic a bit so that the `AlertBanner` takes the whole `bannerAlert` as a prop.

**Review by PaulJKim — APPROVED** [review]
LGTM!
